encoding/protobuf/unaggregated_encoder.go

```diff
@@ -0,0 +1,175 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package protobuf
+
+import (
+	"encoding/binary"
+	"fmt"
+	"math"
+
+	"github.com/m3db/m3metrics/generated/proto/metricpb"
+	"github.com/m3db/m3metrics/metric/unaggregated"
+	"github.com/m3db/m3x/pool"
+)
+
+const (
+	// maximum number of bytes to encode message size.
+	maxMessageSizeInBytes = binary.MaxVarintLen32
+)
+
+// UnaggregatedEncoder encodes unaggregated metrics.
+type UnaggregatedEncoder interface {
+	// Len returns the number of bytes accumulated in the encoder so far.
+	Len() int
+
+	// Reset resets the encoder buffer with optional initial data.
+	Reset(initData []byte)
+
+	// EncodeMessage encodes an unaggregated message.
+	EncodeMessage(msg MessageUnion) error
+
+	// Relinquish relinquishes ownership of the encoded byte stream to the caller,
+	// and resets the internal encoding buffer.
+	Relinquish() Buffer
+}
+
+type unaggregatedEncoder struct {
+	pool           pool.BytesPool
+	initBufSize    int
+	maxMessageSize int
+
+	cm   metricpb.CounterWithMetadatas
+	bm   metricpb.BatchTimerWithMetadatas
+	gm   metricpb.GaugeWithMetadatas
+	buf  []byte
+	used int
+
+	encodeMessageSizeFn func(int)
+	encodeMessageFn     func(metricpb.MetricWithMetadatas) error
+}
+
+// NewUnaggregatedEncoder creates a new unaggregated encoder.
+func NewUnaggregatedEncoder(opts UnaggregatedOptions) UnaggregatedEncoder {
+	e := &unaggregatedEncoder{
+		pool:           opts.BytesPool(),
+		initBufSize:    opts.InitBufferSize(),
+		maxMessageSize: opts.MaxMessageSize(),
+	}
+	e.encodeMessageSizeFn = e.encodeMessageSize
+	e.encodeMessageFn = e.encodeMessage
+	e.Reset(nil)
+	return e
+}
+
+func (enc *unaggregatedEncoder) Len() int { return enc.used }
+
+func (enc *unaggregatedEncoder) Reset(initData []byte) {
+	if enc.buf != nil && enc.pool != nil {
+		enc.pool.Put(enc.buf)
+	}
+	bufSize := int(math.Max(float64(enc.initBufSize), float64(len(initData))))
+	enc.buf = allocate(enc.pool, bufSize)
+	copy(enc.buf, initData)
```

Comment:
How often do you have data in the `initData`? could you just use the initData rather than allocate and copy?

Reply:
Basically if I need to rewind the buffer when the encoded bytes > the flush size, then I need to rewind and copy the extra bytes into a new buffer. I'd prefer to copy so that you can't modify the internal encoder data from outside the encoder.